sklearn/preprocessing.py

```diff
@@ -399,6 +404,120 @@ def __init__(self, copy=True, with_mean=True, with_std=True):
+        `sort_X_`.
+        This is an approximation, to save memory and transform
+        computation time.
+        e.g. if 1000, transformed values will have resolution 0.001.
+        If `None`, we store the full size matrix, comparable
+        in size to the initial fit `X`.
+
+    Attributes
+    ----------
+    `sort_X_` : array of ints, shape (n_samples, n_features)
+        The rank-index of every feature in the fit X.
+
+    See also
+    --------
+    :class:`sklearn.preprocessing.StandardScaler` to perform standardization
+    that is faster, but less robust to outliers.
+    """
+
+    def __init__(self, n_ranks=1000):
+        # TODO: Add min and max parameters? Default = [0, 1]
+        self.n_ranks = n_ranks
+
+    def fit(self, X, y=None):
+        """Compute the feature ranks for later scaling.
+
+        fit will take time O(n_features * n_samples * log(n_samples)),
+        because it must sort the entire matrix.
+
+        It use memory O(n_features * n_ranks).
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The data used to compute feature ranks.
+        """
+        X = array2d(X)
+        n_samples, n_features = X.shape
+        full_sort_X_ = np.sort(X, axis=0)
+        if not self.n_ranks or self.n_ranks >= n_samples:
+            # Store the full matrix
+            self.sort_X_ = full_sort_X_
+        else:
+            # Approximate the stored sort_X_
+            self.sort_X_ = np.zeros((self.n_ranks, n_features))
+            for i in range(self.n_ranks):
+                for j in range(n_features):
+                    # Find the corresponding i in the original ranking
+                    iorig = i * 1. * n_samples / self.n_ranks
+                    ioriglo = int(iorig)
+                    iorighi = ioriglo + 1
+
+                    if ioriglo == n_samples:
+                        self.sort_X_[i, j] = full_sort_X_[ioriglo, j]
+                    else:
+                        # And use linear interpolation to combine the
+                        # original values.
+                        wlo = (1 - (iorig - ioriglo))
+                        whi = (1 - (iorighi - iorig))
+                        assert wlo >= 0 and wlo <= 1
+                        assert whi >= 0 and whi <= 1
+                        assert_almost_equal(wlo+whi, 1.)
+                        self.sort_X_[i, j] = wlo * full_sort_X_[ioriglo, j] \
+                                           + whi * full_sort_X_[iorighi, j]
+        return self
+
+    def transform(self, X):
+        """Perform rank-standardization.
+
+        transform will take O(n_features * n_samples * log(n_ranks)),
+        where `n_fit_samples` is the number of samples used during `fit`.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+        """
+        X = array2d(X)
```

Comment:
I think we should also support sparse data, at least when the data is positive. The quantile transformation would not break the sparsity pattern in that case.